libraries/protocol/custom_authorities/restriction_predicate.hxx

```diff
@@ -0,0 +1,575 @@
+template<typename F, typename C, typename = void> struct predicate_in : predicate_invalid<F, C> {};
+template<typename Field, typename Element>
+struct predicate_in<Field, flat_set<Element>, std::enable_if_t<comparable_types<Field, Element> && !is_safe<Field>>> {
+   // Simple inclusion check
+   constexpr static bool valid = true;
+   bool operator()(const Field& f, const flat_set<Element>& c) const { return c.count(f) != 0; }
+};
+template<typename Field, typename Element>
+struct predicate_in<fc::safe<Field>, flat_set<Element>, std::enable_if_t<comparable_types<Field, Element>>> {
+   // Check for safe value
+   constexpr static bool valid = true;
+   bool operator()(const fc::safe<Field>& f, const flat_set<Element>& c) const { return c.count(f.value) != 0; }
+};
+template<typename Field, typename Element>
+struct predicate_in<fc::optional<Field>, flat_set<Element>, std::enable_if_t<comparable_types<Field, Element>>> {
+   // Check for optional value
+   constexpr static bool valid = true;
+   bool operator()(const fc::optional<Field>& f, const flat_set<Element>& c) const {
+      if (!f.valid()) return predicate_result::Rejection(predicate_result::null_optional);
+       return c.count(*f) != 0;
+   }
+};
+// Field-not-in-list is just field-in-list wrapped in a negator
+template<typename Field, typename Container> struct predicate_not_in : predicate_in<Field, Container> {
+   using base = predicate_in<Field, Container>;
+   bool operator()(const Field& f, const Container& c) const { return !base::operator()(f, c); }
+};
+
+// List-contains-list predicate
+template<typename C1, typename C2, typename = void> struct predicate_has_all : predicate_invalid<C1, C2> {};
+template<typename FieldElement, typename ArgumentElement>
+struct predicate_has_all<flat_set<FieldElement>, flat_set<ArgumentElement>,
+                         std::enable_if_t<comparable_types<FieldElement, ArgumentElement>>> {
+   // Field is already flat_set
+   constexpr static bool valid = true;
+   bool operator()(const flat_set<FieldElement>& f, const flat_set<ArgumentElement>& a) const {
+      if (f.size() < a.size()) return false;
+      return std::includes(f.begin(), f.end(), a.begin(), a.end());
+   }
+};
+template<typename FieldContainer, typename ArgumentElement>
+struct predicate_has_all<FieldContainer, flat_set<ArgumentElement>,
+                         std::enable_if_t<is_container<FieldContainer> && !is_flat_set<FieldContainer> &&
+                                          comparable_types<typename FieldContainer::value_type, ArgumentElement>>> {
+   predicate_has_all<flat_set<typename FieldContainer::value_type>, flat_set<ArgumentElement>> inner;
+   // Field is other container; convert to flat_set
+   constexpr static bool valid = true;
+   bool operator()(const FieldContainer& f, const flat_set<ArgumentElement>& a) const {
+      if (f.size() < a.size()) return false;
+      flat_set<typename FieldContainer::value_type> fs(f.begin(), f.end());
```

Comment:
See below, `flat_set` can be expensive.

Reply:
OK, so what if I take `a` by copy, and iterate the values in `f`, checking if they are in `a`. Each value in `f` I find in `a`, I erase from `a`. Break if `a` goes empty or have checked all values in `f`. At the end, `return a.empty();`

Reply:
Not good. Erasing all elements from a `flat_set` has `O(n^2)` when erasing in the wrong order.

Reply:
So copy `a` to a `std::set` instead?

Reply:
Or copy `f` to a `std::set` and use `std::includes`.

Reply:
Eh, way simpler. OK, will do.